kvserver: add metric for Raft leader removals

erikgrinaker opened the pull request:
This patch adds the metric `range.raftleaderremovals` which counts the number of times a Raft leader was removed from a range via a config change. Other removals, such as range merges, are excluded.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/104320)
<!-- Reviewable:end -->

tbg commented on pkg/kv/kvserver/metrics.go:
```diff
@@ -1009,6 +1009,12 @@ var (
 	}
+	metaRangeRaftLeaderRemovals = metric.Metadata{
+		Name:        "range.raftleaderremovals",
+		Help:        "Number of times the current leader was removed from a range (per replica)",
```
"raft leader"

tbg reviewed (changes requested):
Looks good, just need to sort out where to increment the metric. The desired effect should be that the counter "usually" exactly tracks the actual number of times we've removed a leader, which I doubt the code currently does.

erikgrinaker commented on pkg/kv/kvserver/replica_application_result.go:
```diff
@@ -418,6 +419,10 @@ func (r *Replica) handleChangeReplicasResult(
 	}
 
+	if r.raftBasicStatusRLocked().RaftState == raft.StateLeader {
+		r.store.metrics.RangeRaftLeaderRemovals.Inc(1)
```
This is noted in the commit message:

>The metric is counted separately per replica, since different replicas can have a different view of the Raft leader.

Specifically, I'm interested in whenever a replica thinks the leader was removed from the range, even though the actual leader may currently be someone else.

The motivation here is to know when various code paths that depend on this get triggered, which happens independently on each replica.

erikgrinaker replied:
On second thought, I think you're right, it only makes sense to count this once. Changed it to count when the leader removes itself. This code is run before the config change is applied to Raft, so it will still see itself as the leader (assuming it still is the leader).

erikgrinaker committed:
kvserver: add metric for Raft leader removals

This patch adds the metric `range.raftleaderremovals` which counts the
number of times a Raft leader was removed from a range via a config
change. Other removals, such as range merges, are excluded.

Epic: none
Release note: None

erikgrinaker commented:
bors r+
